src/Microsoft.CorrelationVector/CorrelationVector.cs

```diff
@@ -1,397 +1,122 @@
-        /// <param name="vector">
-        /// The correlation vector you want to compare with the current correlation vector.
-        /// </param>
-        /// <returns>
-        /// True if the specified correlation vector is equal to the current correlation
-        /// vector; otherwise, false.
-        /// </returns>
-        public bool Equals(CorrelationVector vector)
-        {
-            return string.Equals(this.Value, vector.Value, StringComparison.Ordinal);
-        }
-
-        private CorrelationVector(string baseVector, int extension, CorrelationVectorVersion version, bool immutable)
-        {
-            this.BaseVector = baseVector;
-            this.extension = extension;
-            this.Version = version;
-            this.immutable = immutable || CorrelationVector.IsOversized(baseVector, extension, version);
-        }
-
-        private static string GetUniqueValue(CorrelationVectorVersion version)
-        {
-            if (CorrelationVectorVersion.V1 == version)
-            {
-                byte[] bytes = Guid.NewGuid().ToByteArray();
-                return Convert.ToBase64String(bytes, 0, 12);
-            }
-            else if (CorrelationVectorVersion.V2 == version)
-            {
-                return Guid.NewGuid().GetBaseFromGuid();
-            }
-            else
-            {
-                throw new ArgumentException(string.Format(CultureInfo.CurrentCulture, "Unsupported correlation vector version: {0}", version));
-            }
-        }
-
-        private static CorrelationVectorVersion InferVersion(string correlationVector, bool reportErrors)
-        {
-            int index = correlationVector == null ? -1 : correlationVector.IndexOf('.');
-
-            if (CorrelationVector.BaseLength == index)
-            {
-                return CorrelationVectorVersion.V1;
-            }
-            else if (CorrelationVector.BaseLengthV2 == index)
-            {
-                return CorrelationVectorVersion.V2;
-            }
-            else
-            {
-                //By default not reporting error, just return V1
-                return CorrelationVectorVersion.V1;
-            }
-        }
-
-        private static bool IsImmutable(string correlationVector)
+                //By default not reporting error, just return V1
+                return CorrelationVectorVersion.V1;
+            }
+        }
+
+        /// <summary>
+        /// Converts a string representation of a Correlation Vector into this class.
+        /// </summary>
+        /// <param name="correlationVector">String representation</param>
+        /// <returns>The Correlation Vector based on its version.</returns>
+        public static CorrelationVector Parse(string correlationVector)
+        {
+            CorrelationVectorVersion version = CorrelationVector.InferVersion(correlationVector);
+            return RunStaticMethod(correlationVector, version, CorrelationVectorV1.Parse, CorrelationVectorV2.Parse);

```

Comment:
I would probably recommend the same approach of using a switch statement (like how we did in Spin) instead of this RunStaticMethod class.

Reply:
I've rewritten RunStaticMethod in order to handle any number of versions because I don't want to see too many switch statements.

Reply:
Right but you're going to have to keep on changing the function definition for the RunStaticMethod so just adding another case in a switch statement might be more convenient.